WIP: Implement tracing of a command that is provided as additional parameters

landesfeind opened the pull request:
I propose the following changes to enable tracing a command that is provided on command line instead of a PID. This enable the use of `procrec` similar to the Linux `time` command.

I would be happy to receive some feedback and incorporate more suggestions if you have :)


## Example usage
```sh 
$ ./procrec -i 1 sleep 3
0.00 PID 18823 CPU% 0.00 RSS 606 VSIZE 8613 
1.00 PID 18823 CPU% 0.00 RSS 606 VSIZE 8613
```

I adjusted the help/usage text like such:
```
$ ./procrec -h
procrec 0.2.2
Lars Baumgaertner
Process recorder to log cpu utilization and memory consumption

USAGE:
    procrec [FLAGS] [OPTIONS] [command]...

ARGS:
    <command>...    The command to execute and record. If omitted, then --pid must be provided

FLAGS:
    -g, --graph            Display graph using gnuplot
    -h, --help             Prints help information
    -t, --print-gnuplot    Just print gnuplot script
    -v, --verbose          A level of verbosity, and can be used multiple times
    -V, --version          Prints version information

OPTIONS:
    -d, --duration <duration>    Duration for observation
    -i, --interval <interval>    Sampling interval in seconds [default: 2]
    -p, --pid <pid>              Process to be inspected. If omitted, then command must be provided
```

## Example of error messages upon false invocation:
```
./procrec 
error: The following required arguments were not provided:
    --pid <pid>
```
This is a bit unfortunate but I am not sure why 
```
./procrec --pid 1 sleep 1
error: The argument '<command>...' cannot be used with '<command>...'
```
An alternative would be to simply have `--pid` overrule `command` inside the program logic and comment on this in the help text?

gh0st42 commented:
Thank you for this useful PR! 
This feature was also on my wishlist :) 

It would be really nice if you could find a solution to the short error message so that it includes also the command alternative:
```
$ procrec
error: The following required arguments were not provided:
    --pid <pid>

USAGE:
    procrec --interval <interval> --pid <pid>

For more information try --help
```

Otherwise, this is pretty confusing for users. Did not look into clap for a while but there must be a way to include both valid invocation methods in the help text.

gh0st42 commented on src/main.rs:
```diff
@@ -128,33 +154,51 @@ fn main() {
         }
     }
 
+    match child {
```
using `if let Some(mut c) = child` would be more idiomatic and doesn't need the empty none branch

gh0st42 commented on src/main.rs:
```diff
@@ -128,33 +154,51 @@ fn main() {
-        if opts.verbose > 0 {
-            println!("{}", data);
+        match child.as_mut() {
+          Some(c) => match c.try_wait() {
```
instead of these double nested match statements and since we are only interested in the `Some(Ok(Some(_)))` case and in case of a failure you are panicing anyhow it would probably be nicer to `expect` instead of having the `Err` branch and make use of `if let` for readability and since we are only interested in the final `Some` case.

gh0st42 replied:
maybe also a comment why this is necessary, why is not the later `pid_proc.is_running()` enough to check if the process is accessible and running?

gh0st42 commented on src/main.rs:
```diff
@@ -110,9 +115,30 @@ fn main() {
+        let mut cmd = Command::new(cl.remove(0));
+        cmd.args(cl);
+
+        let pid;
```
this does not seem very idiomatic, any other way you can express that?

gh0st42 commented on src/main.rs:
```diff
@@ -110,9 +115,30 @@ fn main() {
         println!("{}", gnuplot_script_content);
         std::process::exit(0);
     }
+    let mut child : Option<std::process::Child> = None;
```
some comment would be nice why this is needed and in which case

gh0st42 reviewed (changes requested):
Some changes to the code would be nice to keep it maintainable in the future.

Thank you for your work!

landesfeind commented:
> Thank you for this useful PR!
> This feature was also on my wishlist :)

Glad to hear! :) If you find it useful, I will put some effort in tidying up the code and improve (thanks a lot for the suggestions!!). Might take me a few days, but I will get back to you :)

landesfeind commented:
Please check the updated version. I followed you recommendation for a more ideomatic way and also added more comments. To further ease the understanding and reduce the burden of having to track two variables (i.e., `pid_proc` and `child`) I encapsulated them in a new `enum TrackedProcess`.

For the error message: unfortunately, there does not seem to be a "one of these paramter needs to be present" (at least not in the current beta version of clap). I now implemented a custom error message.

Edit: here is a compilation of error messages:
```
+ ./procrec
Error: Process to record must be provided as additional argument or via '--pid' parameter. For detailed information, execute with --help

+ ./procrec hostname
my-hostname

+ ./procrec --pid 1 --duration 1
0.00 PID 1 CPU% 0.00 RSS 12132 VSIZE 171855 
2.00 PID 1 CPU% 0.00 RSS 12132 VSIZE 171855 

+ ./procrec --pid 1 --duration 1 hostname
error: The argument '<command>...' cannot be used with '--pid <pid>'

USAGE:
    procrec --pid <pid> --duration <duration> --interval <interval>

For more information try --help
```


PS: Sorry for some not-so-nice indentations. I will apply `cargo fmt` after your review because it will introduce changes to many additional lines which make the review harder.

gh0st42 commented on src/main.rs:
```diff
@@ -72,11 +77,114 @@ impl fmt::Display for Sample {
+    type Error = String;
+
+    fn try_from(opts: &'a Opts) -> Result<Self, Self::Error> {
+     match opts.pid {
```
I'm still not sure if these nested matches are the most idiomatic and maintainable way.
another option would be something like `opts.unwrap_or_else(|e| panic!("err: {}", e))` or when you are directly returning a result again just `map` it from one `Ok` to another.
If the error handling gets cumbersome adding `anyhow` crate could be an option, even though I'd like to keep the required deps to a minimum.

gh0st42 commented on src/main.rs:
```diff
@@ -72,11 +77,114 @@ impl fmt::Display for Sample {
+         }
+       
+         match cmd.spawn() {
+           Ok(c) => match Process::new(c.id()) {
```
Same as above:
I'm still not sure if these nested matches are the most idiomatic and maintainable way.
another option would be something like `opts.unwrap_or_else(|e| panic!("err: {}", e))` or when you are directly returning a result again just `map` it from one `Ok` to another.
If the error handling gets cumbersome adding `anyhow` crate could be an option, even though I'd like to keep the required deps to a minimum.

gh0st42 commented on src/main.rs:
```diff
@@ -111,8 +220,16 @@ fn main() {
-    let mut pid_proc = Process::new(opts.pid).expect("Failed accessing process");
+    // Initialize the tracking process
+    let mut pid_proc = match TrackedProcess::try_from(&opts) {
+			Err(e) => { 
```
Another way to shorten the program would be to make more use of the `?` operator for error handling as in most cases either just the error is printed or the function/program is exited

gh0st42 replied:
Especially when also using `anyhow` lots of `unwraps()` can also be replaced

gh0st42 reviewed (changes requested):
I really like that you took early exiting of `procrec` into account and cleanly killing the spawned process.

The program got quite a bit more complex now for such a simple task.
Maybe bubbling up errors more, mapping them when unwrapping and use of the `?` operator can simplify the code again. I believe the `anyhow` crate only introduces 2-3 dependencies and compiles pretty quickly. Might be worth the effort.
Also regarding the nested matches some other options are mentioned here: https://stackoverflow.com/questions/38522870/are-nested-matches-a-bad-practice-in-idiomatic-rust

Too bad that `clap` does not have an easy solution for the exclusive parameters, maybe that will change or a feature request is needed.

Your comments in the source are great and will probably help in the long run!
Thank you anyhow for your great work so far!

landesfeind commented:
> The program got quite a bit more complex now for such a simple task.

Well, I guess that is simply whats happening if you expand the functionality :D On the other hand - I think the complexity is nicly encapsulated in the new struct.

Regarding nested matching and idiomatic rust: I don't know - I feel that matches make it quite idiomatic because they ensure that you don't miss out on a particular enum value etc. To me it feels quite natural...  

You are welcome to use the code and also introduce more additional dependencies like anyhow - the proc currently does perfectly well what I need it do. So - thanks a lot :)
